### PR TITLE
GODRIVER-2753 Integration with oss-fuzz fuzzing service

### PR DESCRIPTION
**Pull request by manunio**
GODRIVER-2753

## Summary
This PR adds a cmd to build zip file of seed corpus for fuzzing to be used by oss-fuzz.

## Background & Motivation
Hi, I would like to help integrate this project into [OSS-Fuzz](https://github.com/google/oss-fuzz).

- As an initial step for integration I have created this PR: https://github.com/google/oss-fuzz/pull/9528, it contains necessary logic from an OSS-Fuzz perspective to integrate mongo-go-driver.

- OSS-Fuzz is a free service run by Google that performs continuous fuzzing of important open source projects, you can find more details in our [FAQ](https://google.github.io/oss-fuzz/faq/).

-  If you would like to integrate, the only thing I need is a list of email(s), it must be associated with a google account like gmail ([why?](https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication)). by doing that, the provided email(s) will get access to the data produced by OSS-Fuzz, such as bug reports, coverage reports and more stats.

- Notice the email(s) affiliated with the project will be public in the OSS-Fuzz repo, as they will be part of a configuration file.

### CODE REVIEW

File: bson/bson_fuzz_corpus_common.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+package bson
+
+import (
+	"fmt"
+	"os"
+	"path"
+)
+
+type TestCase struct {
```
Comment (prestonvasquez) [inline bson/bson_fuzz_corpus_common.go:15]: There is no need for this common file or to export any of this data. Can we unexport it and just duplicate the "findJSONFilesInDir" and "JsonToBytes" functions in the cmd/build-oss-fuzz-corpus/main.go ?

File: cmd/build-oss-fuzz-corpus/main.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+		seedTestCase(zw, &tcase)
+	}
+}
+
+// This cmd generates and adds slice of bytes to files with sha1 hash name, and zip them as a <fuzzer_name>_seed_corpus.zip,
+// which is used later by oss-fuzz as seed corpus, This is done because as of now oss-fuzz does not support go native t.Add()
+// method.
+func main() {
+	seed_corpus := os.Args[1]
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:98]: Rename this to "seedCorpus", [mixedCaps](https://go.dev/doc/effective_go#mixed-caps) is the convention.

File: cmd/build-oss-fuzz-corpus/main.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	if !strings.HasSuffix(seed_corpus, ".zip") {
+		log.Fatalln("Expected command line:", os.Args[0], "<seed_corpus>.zip")
+	}
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:101]: What is this checking for? If it's to see if the corpus has a zip extension can you do it this way:

```suggestion
	if filepath.Ext(seedCorpus) != ".zip" {
		log.Fatalf("expected zip file <corpus>.zip, got %s", seedCorpus)
	}
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+	zip_file, err := os.Create(seed_corpus)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:103]: Rename this to "zipFile", [mixedCaps](https://go.dev/doc/effective_go#mixed-caps) is the convention.

File: cmd/build-oss-fuzz-corpus/main.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		log.Fatalf("Failed creating file: %s", err)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:105]:
```suggestion
		log.Fatalf("failed to create zip file: %v", err)
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	defer func() {
+		err := zip_file.Close()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	zip_writer := zip.NewWriter(zip_file)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:115]: Rename this to "zipWriter", [mixedCaps](https://go.dev/doc/effective_go#mixed-caps) is the convention.

File: cmd/build-oss-fuzz-corpus/main.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	seedBSONCorpus(zip_writer)
+
+	if err := zip_writer.Close(); err != nil {
+		log.Fatal(err)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:119]:
```suggestion
		log.Fatalf("failed to close zip writer: %v", err)
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+			log.Fatal(err)
+		}
+
+		seedTestCase(zw, &tcase)
+	}
+}
+
+// This cmd generates and adds slice of bytes to files with sha1 hash name, and zip them as a <fuzzer_name>_seed_corpus.zip,
+// which is used later by oss-fuzz as seed corpus, This is done because as of now oss-fuzz does not support go native t.Add()
+// method.
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:96]:
```suggestion
// main packs the local corpus as <fuzzer_name>_seed_corpus.zip, which is used by OSS-Fuzz to seed remote fuzzing of the MongoDB Go Driver. See here for more details: https://google.github.io/oss-fuzz/architecture/ 
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		filePath := path.Join(dataDir, fileName)
+
+		file, err := os.Open(filePath)
+		if err != nil {
+			log.Fatalf("failed to open file %q: %v", filePath, err)
+		}
+
+		var tcase bson.TestCase
+		if err := json.NewDecoder(file).Decode(&tcase); err != nil {
+			log.Fatal(err)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:87]:
```suggestion
			log.Fatalf("failed to decode file %q: %v", filePath, err)
```

**Review by prestonvasquez — CHANGES_REQUESTED** [review]
Thank you @manunio for putting this together, the team is very appreciative. I have some initial feedback, please reach out if you have any questions. 

As for the emails,

- "primary_contact": dbx-go@mongodb.com
- "auto_cc": 
    - qingyang.hu@mongodb.com
    - matt.dale@mongodb.com
    - kevin.albertson@mongodb.com
    - preston.vasquez@mongodb.com

File: cmd/build-oss-fuzz-corpus/main.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+type decodeErrorTestCase struct {
+	Description string `json:"description"`
+	Bson        string `json:"bson"`
+}
+
+type parseErrorTestCase struct {
+	Description string `json:"description"`
+	String      string `json:"string"`
+}
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:54]:
```suggestion
type validityTestCase struct {
	Description       string  `json:"description"`
	CanonicalExtJSON  string  `json:"canonical_extjson"`
	RelaxedExtJSON    *string `json:"relaxed_extjson"`
	DegenerateExtJSON *string `json:"degenerate_extjson"`
	ConvertedExtJSON  *string `json:"converted_extjson"`
}
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		if err != nil {
+			log.Fatalf("failed to open file %q: %v", filePath, err)
+		}
+
+		var tcase testCase
+		if err := json.NewDecoder(file).Decode(&tcase); err != nil {
+			log.Fatalf("failed to decode file %q: %v", filePath, err)
+		}
+
+		seedTestCase(zw, &tcase)
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:163]:
```suggestion
		tc := struct {
			Valid []*validityTestCase `json:"valid"`
		}{}

		if err := json.NewDecoder(file).Decode(&tc); err != nil {
			log.Fatalf("failed to decode file %q: %v", filePath, err)
		}

		seedTestCase(zw, tc.Valid)
```

File: cmd/build-oss-fuzz-corpus/main.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+	_, err = f.Write(jbytes)
+	if err != nil {
+		log.Fatalf("failed to write file: %s into zip file: %v", zipFile, err)
+	}
+}
+
+// seedTestCase will add the byte representation for each "extJSON" string of each valid test case to the fuzzer's
+// corpus.
+func seedTestCase(zw *zip.Writer, tcase *testCase) {
+	for _, vtc := range tcase.Valid {
```
Comment (prestonvasquez) [inline cmd/build-oss-fuzz-corpus/main.go:122]:
```suggestion
func seedTestCase(zw *zip.Writer, tcase []*validityTestCase) {
	for _, vtc := range tcase {
```

**Review by prestonvasquez — CHANGES_REQUESTED** [review]
@manunio We don't really need the "testCase", "decodeErrorTestCase", or "parseErrorTestCase" structs. We just need a subset of the "validityTestCase" struct, I've updated the PR with some suggestions on how to refactor to remove these structs.

**Review by prestonvasquez — APPROVED** [review]

**Review by matthewdale — CHANGES_REQUESTED** [review]
One required change to support compiling with Go 1.13, which is a currently a requirement for all non-test code in the `mongo-go-driver` repo. Otherwise, looks great!

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Comment by matthewdale** [comment]
@manunio I realized after merging this that the "gosec" linter complains about the `"crypto/sha1"` package because of "weak cryptographic primitive". Does the oss-fuzz integration rely on using SHA1 hashes for filenames, or can we update it to use SHA256 hashes?

**Pull request merged**